src/renderer/utils/subscriptions-manager.ts

```diff
@@ -0,0 +1,205 @@
+// SPDX-FileCopyrightText: 2025 The BAR Lobby Authors
+//
+// SPDX-License-Identifier: MIT
+
+import { UserId } from "tachyon-protocol/types";
+import { Signal } from "$/jaz-ts-utils/signal";
+
+export class SubsManager {
+    public readonly onUsersAttached: Signal<UserId[]> = new Signal();
+    public readonly onUsersDetached: Signal<UserId[]> = new Signal();
+
+    private userList: Map<UserId, Set<symbol>>;
+    constructor() {
+        this.userList = new Map<UserId, Set<symbol>>();
+    }
+    /**
+     * Attach the UserId(s) to the selected subscription list. A subscription request will be automatically sent to the Tachyon server if required.
+     * @param users UserId, or Array of UserId, that will be attached
+     * @param list Symbol indicating which subscription list to attach them to.
+     */
+    attach(users: UserId, list: symbol): void;
+    attach(users: UserId[], list: symbol): void;
+    attach(users: UserId | UserId[], list: symbol): void {
+        const userSubs: UserId[] = [];
+        for (const user of typeof users === "string" ? [users] : users) {
+            //for every userId we will...
+            let dirty: boolean = false;
+            let uLists = this.userList.get(user); //get the lists for the user
+            if (!uLists) {
+                //if undefined, there are none for this user
+                uLists = new Set(); //so we create an empty Set for this user
+                uLists.add(list); //we add the requested list to this set
+                userSubs.push(user); //add the user to the subscription list because they are entirely new
+                dirty = true;
+            }
+            if (!uLists.has(list)) {
+                //if this list is missing from this user
+                uLists.add(list); //we add it!
+                dirty = true;
+            }
+            if (dirty) {
+                this.userList.set(user, uLists); //update the user's map with the changes
```

Comment:
do it when you create new `ulist` instead separately, no need to split it and introduce `dirty` bit.